diesel_derives/src/selectable.rs

```diff
@@ -136,9 +136,25 @@ fn to_field_ty_bound(field_ty: &syn::Type) -> Option<TokenStream> {
 }
 
 fn field_column_ty(field: &Field, model: &Model) -> Result<TokenStream> {
-    if let Some(ref select_expression_type) = field.select_expression_type {
+    if let Some(ref select_expression) = field.select_expression {
+        use dsl_auto_type::auto_type::expression_type_inference as type_inference;
+        let expr = &select_expression.item;
+        let (inferred_type, errors) = type_inference::infer_expression_type(
+            expr,
+            field.select_expression_type.as_ref().map(|t| &t.item),
+            &type_inference::InferrerSettings::builder()
+                .dsl_path(parse_quote!(diesel::dsl))
+                .function_types_case(crate::AUTO_TYPE_DEFAULT_FUNCTION_TYPE_CASE)
+                .method_types_case(crate::AUTO_TYPE_DEFAULT_METHOD_TYPE_CASE)
+                .build(),
+        );
+        let mut type_token_stream = quote!(#inferred_type);
```

Comment:
Can we use `quote_spanned!` here with a span of the original expression? That should result in the compiler pointing at the `#[diesel(select_expression = …)]` type in the relevant error message. (Also would be great to have a compile test for that)

Reply:
Actually quote_spanned only puts the span on the tokens that are directly written in the macro, but `#inferred_type` just keeps its span.
There was however an issue with the generated `_` placeholders though, which were not ideally spanned, so that is fixed in the type inferrence code.